### DISCUSSION TITLE
🎉 New Release: Item Menus, No Wrap, Link to Folders (v1.9)

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#97] [Announcements]
This release includes the ability to add menus (and sub-menus), disable wrapping on mobile, link to folders, and see toolbar usage. 

_I will be exploring adding other sub-toolbar types in upcoming releases. See the [updated roadmap](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Roadmap)._

## Installation

Install via Community Plugins.

# New Features 🎉

- Adds an `Item Menu` item type to the available toolbar items:
  - After selecting the `Item Menu` item type, select an existing toolbar, and that item will display the toolbar's items in a dropdown menu when triggered.
  - Sub-menus are also supported.
  - Menus can be styled via the [Style Settings plugin](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Style-Settings-plugin-support), or by CSS snippets (via the `.note-toolbar-menu` class).
  - Menu items are fully keyboard accessible (via the `Focus` command). Press `Enter` to open a menu item, and ⬆️⬇️ to navigate.
  - Note that menus in Obsidian on mobile appear at the bottom of the screen when opened.
  
![example_item_menu](https://github.com/chrisgurney/obsidian-note-toolbar/blob/master/docs/example_item_menu.png?raw=1)

- Adds a `don't wrap items` (`mnwrp`) Style specifically for mobile. When set, all items will be on one line making the toolbar scrollable.
  - By default (i.e., when it's not set), items wrap to the next line.
  - For it to work best, ensure your items are not centered or evenly spaced (i.e., left-align them).

<img src="https://github.com/user-attachments/assets/12614b63-6870-4002-bde2-231758ec788c" title="Edit Toolbar dialog, showing name, items, styles, and a delete button" width="500"/>

- You can now link to folders from a `File` type. When clicked, the folder is highlighted in the File Explorer (behaves the same way as if the folder was bookmarked) • _Thank you, indirect [Reddit](https://www.reddit.com/r/ObsidianMD/comments/1drtd11/is_there_any_way_to_focus_on_folders_or_have_a/) [suggestion](https://imgur.com/a/HcJKZLc)!_
- A _Usage_ section has been added to a toolbar's Settings (below the _Styles_ section), where you can see the number of times the toolbar is used (in mappings, and items). A link is also provided to search for toolbar properties that might be using this toolbar.

![CleanShot 2024-07-08 at 12 53 53@2x](https://github.com/chrisgurney/obsidian-note-toolbar/assets/17337703/25d7833a-d4c1-4a62-9796-3eb2bfdff3fc)

# Beta Known Issues 🐞

## Menus

1. ~~Labels aren't showing in items, if set.~~ _My bad; non-issue._
2. ~~There's a border missing in the Settings toolbar list section.~~ _Fixed in `beta-2`_
3. ~~Menu appears under label when label is clicked (vs. left edge / icon)~~ _Fixed in `beta-3`_
4. ~~Sub-menus not positioned correctly:~~ _Fixed in `beta-5`_
    a. ~~Desktop: There's a gap between sub-menus and the menu that's triggered (desktop).~~
    b. ~~Tablet only: Sub-menus opened from a FAB menu open in the lower left when using keyboard, and below the menu when tapping.~~
    c. ~~Click, then use keyboard to open a sub-menu → It opens in the wrong position.~~
5. ~~Settings: When toolbar previews are longer than the ToolbarSuggester field, it moves down to the next line.~~ _Fixed_

**Comment by FelipeRearden** [comment]
Amazing new features @chrisgurney !!! Looks a strong first beta release.

I am gonna be a Tester for the betas of version `1.9.0` for Obsidian Mobile on iPadOS with keyboard.

> Menu items are fully keyboard accessible (via the Focus command).

Thanks for adding support for keyboard navigation 💙

**Reply by chrisgurney** [reply]
Appreciate the help with testing, as always!

**Reply by FelipeRearden** [reply]
@chrisgurney , just a question before dive into the beta…
1️⃣ Is there a specific CSS rules to target ONLY the items inside de menu? 

2️⃣ If yes, these rules are specific for the NTB plugin or are you using the same rules from Obsidian standard menus (like the context menu for example)

**Reply by chrisgurney** [reply]
@FelipeRearden It's a standard Obsidian menu, with no special CSS applied. 

I had thought about adding a decoration to any items that are sub-menus, but I think I'll leave that up to the user for now (can add something to the label if they wish).

**Reply by chrisgurney** [reply]
@FelipeRearden Sorry I'm realizing now that this was a general question:

> Is there a specific CSS rules to target ONLY the items inside de menu?

I'm sure you've figured it out by now but`div.menu` is right below the `body` (I know you're not on desktop, but for anybody else reading this it probably only applies if using non-native menus):

![CleanShot 2024-07-12 at 09 46 31@2x](https://github.com/user-attachments/assets/04e3946d-e673-47f3-bfa0-8ee3e1a229bb)

**Comment by FelipeRearden** [comment]
Feedback - version `1.9.0 beta 2`

The new Menus is working very well in general. 

From the Menu/Submenu keyboard navigation, I have two comments to share:

## 1️⃣ **BR? Focus on the First Menu Item**

After hitting <kbd>ENTER</kbd> to Open the Menu the first item in the Menu is not focused. Had to hit <kbd> ↓</kbd> to focus on the first .

What is driving em crazy is that it was not doing this yesterday (same version from today). I excluded the NTB and started it over and get the same result from above.

Let see the behavior on you end.

<hr>

## 2️⃣ **IDEA: New Feature to  make the MENU keyboard navigation more FAST**

When I think about Menus, would be a great QoL with we dont have to hit <kbd>ENTER</kbd> to open the Menu. Since  using <kbd> ↓</kbd> to open a Menu is not a good option because it would hurt the global keyboard navigation, I got this idea:

When the NTB item that is focused in a MENU if the item is focused for more the 2000ms it will automatically open the menu.

In other words, a keyboard event listener to perform n action if there is no keyboard event in the next 2000ms.
- Obsidian in able to execute this. I have templater scripts that uses keyboard events listeners (I cant believed that worked 🤪). 

And of course this would be a toggle on Settings to make everybody happy.

**Reply by chrisgurney** [reply]
@FelipeRearden Thanks for your thoughts!

Re: 1️⃣ I'm not exactly sure how to select the first item in the menu, aside from possibly emulating an additional keystroke. I'll give that a go when I get a chance, but I believe it will have to be behind a setting (per the below).

Re: 2️⃣ I think you're right about a toggle that you can switch on to automatically open menus. When it comes to accessibility in general, I'm reading things [like this](https://www.a11y-collective.com/code/mastering-web-accessibility-making-drop-down-menus-user-friendly/#:~:text=submenus%20should%20not%20open%20automatically%20when%20tabbing%20through%20the%20main%20menu%20items.%20Instead,%20the%20submenu%20should%20only%20open%20when%20the%20user%20chooses%20to%20do%20so.) about keyboard navigation and drop-downs:

> A common issue with keyboard navigation in drop-down menus is the need to tab through every single submenu item to reach the next top-level menu item. This can be time-consuming and frustrating for users.

> To prevent this, submenus should not open automatically when tabbing through the main menu items. Instead, the submenu should only open when the user chooses to do so.

**Reply by FelipeRearden** [reply]
> Re: 1️⃣ I'm not exactly sure how to select the first item in the menu, aside from possibly emulating an additional keystroke. I'll give that a go when I get a chance, but I believe it will have to be behind a setting (per the below).

The funny thing is I am almost sure that it was doing this by default when I started my tests. I am gonna redo this from the zero installing NTB again to see if I get some pattern. The obsidian menu behavior is to select the first item.

I let you know if I get some pattern to demonstrate the behavior.

> Re: 2️⃣ I think you're right about a toggle that you can switch on to automatically open menus.

I know that I open NTB many times per hour and another ENTER to open the menu will be hard to apply. I am happy to know that this type of issue has been discussed between experts. We need the two ways because we have users with different workflows for different reasons 💙

**Reply by chrisgurney** [reply]
Further reading about [accessibility of toolbars](https://www.w3.org/WAI/ARIA/apg/patterns/toolbar/).

Based on the [example](https://www.w3.org/WAI/ARIA/apg/patterns/toolbar/examples/toolbar/#ex_label), the suggestion seems to be that the `↓` down arrow opens the menu and focusses the first element in the menu (try it in the font menu, for example).

Perhaps we reserve `↑` up and `↓` down arrow keys for menu navigation.

**Reply by FelipeRearden** [reply]
> Perhaps we reserve ↑ up and ↓ down arrow keys for menu navigation.

Let me see if we are on the same page…   this means that I will not be able to use ↓ to switch between items inside the NTB anymore?

**Reply by chrisgurney** [reply]
> this means that I will not be able to use ↓ to switch between items inside the NTB anymore?

Possibly. I need to review accessibility best practices in detail on this point.

What are your thoughts on having to use left/right instead, and using up/down within menus?

**Reply by FelipeRearden** [reply]
> What are your thoughts on having to use left/right instead, and using up/down within menus?

2 Reasons (which are only personal reasons)…

1️⃣ After a lot of changes in the hotkeys I could let <kbd>CMD ↓</kbd> to Focus a NTB. Since my right hand is already placed in the arrow <kbd>↓</kbd>. I navigate inside the NTB easily this way. 

2️⃣ Due to a complex workflow that I have in vault related to text replacement for callout contents, I have to tweak my `left/right` arrows keys to perform hidden actions + go to right/left. The effect of this on NTB is that **I lost the left/right navigation** inside NTB for these keys.  
> Just an example, I cant edit the YAML using the keyboard. That’s is reason for that HR in #96 . I changed all yaml edit to templater to use the keyboard and avoid open the YAML container.

<hr>

That’s is the reason that I am scared about ↓ down arrow opens the menu. If I lost the ↓, I would lost my entire NTB navigation. 

As I said above, all personal reasons. But I kindly ask if you decide go on this approach (instead of automatic open after 2000ms) if you could accommodate this issue in toggle.
- I rather hit ENTER to open the Menu than lost the entire navigation 🙏

**Reply by chrisgurney** [reply]
@FelipeRearden Appreciate the explanation. 

I can't guarantee the behavior of the up/down keys won't change in the future, but for now I've _only_ updated the behavior after pressing `Enter` so that it focusses within the menu by emulating a down arrow keystroke. There's a small caveat with settings however: see the notes for [1.9.0-beta-3](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.9.0-beta-3)

LMK how it works for you!

**Reply by FelipeRearden** [reply]
> I can't guarantee the behavior of the up/down keys won't change in the future

100% understood and thank you very much for considering, for now, my heterodox workflow 🙏

> I've just updated the behavior after pressing Enter so that it focusses within the menu by emulating a down arrow keystroke

Tested in version `1.9.0-beta` and it is working perfect ! I tested all scenarios that I could image, combining Single Items with Menu and even a NTB with only Menu and it is perfect!

> There's a small caveat with settings however: see the notes for [1.9.0-beta-3](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.9.0-beta-3)
>
>> Note that this only works if the Obsidian setting Appearance > Native menus is off.

100% understood. Just to let you know that on Obsidian Mobile, we DONT HAVE this toggle.  

<hr> 

I am gonna keep stressing `beta 3` and I let you know if I get something. I am gonna try to deal with the CSS inside a Menu (my first attempt was  a nightmare)

Thanks again for everything @chrisgurney ☀️

**Reply by chrisgurney** [reply]
@FelipeRearden Great to hear!

I've updated the release notes with your note about not needing the setting on mobile. I also noted that this change does not work in `Floating button` position toolbars.

What exactly are you attempting to do with the CSS within the menu? I'm curious.

**Reply by FelipeRearden** [reply]
> What exactly are you attempting to do with the CSS within the menu? I'm curious.

Remember that I am a newbie, okay. What is a nightmare are for me could be nothing to experts like you :)

My `.menu` , `.menu-item` and `svg` are customizable (I have a CSS snippet). In other words, I know how to target `.menu` and the elements inside inside menu like `.menu-item` and the icon (svg)

Suppose my `.menu background-color is yellow `and `border is 5px solid red`

My Nightmare: for the NTB menu I need to change the background-color to red, the border to none and change the width + height of the icons and their colors.  

I fail to be able to target `.menu` ONLY FOR the NTB because I can't find how to connect using `:has` `:is` or any other selector.

Maybe something triggers on your mind 🙏

**Reply by chrisgurney** [reply]
@FelipeRearden Interesting! You're making me think that would be nice if I could provide Style Settings options for menus.

Thinking thus far:

- There's no documented way for me to assign a class to the menu, which I could easily do if the API to get its element was available. I've had to work around the Obsidian API before, so this might be worth exploring a little further.
- Assigning a class to the body element would not help, as it wouldn't let you specifically target menus created by Note Toolbar.
- Listening to the DOM to wait for a menu to appear after it's created wouldn't be reliable, I think.
- I've been reading about proxies, and this doesn't seem to help me either, due to no helpful APIs being available.

**Reply by chrisgurney** [reply]
@FelipeRearden Good news!

Thanks to @pjeby who shared how to [get access to the menu's DOM](https://github.com/pjeby/quick-explorer/blob/34d1c868b4c88672caecbfad060cd02da4f13649/src/menus.ts#L10). It's an undocumented API, but it only took like six lines of isolated code for me to add.

I'll be adding the class `note-toolbar-menu` to all toolbar menus in the next beta, and thinking through what to add to Style Settings! Thanks again for sharing your use case; I'm glad I asked!

**Reply by FelipeRearden** [reply]
Wooowwww, this amazing news !!!!!! 
Thank you for sharing this with me.

OMG be able to style the menus will be one more game changer feature for Note Toolbar.
What a great way to start the weekend 🚀

**Reply by chrisgurney** [reply]
Here's all the CSS variables that will be available shortly:
```css
.note-toolbar-menu {
    background-color: var(--cg-nt-menu-bg-color);
    border-color: var(--cg-nt-menu-border-color);
    border-radius: var(--cg-nt-menu-border-radius);
    & .menu-item {
        border-radius: var(--cg-nt-menu-item-border-radius);
        color: var(--cg-nt-menu-item-text-color);
        font-size: var(--cg-nt-menu-item-font-size);
        & .menu-item-icon {
            color: var(--cg-nt-menu-icon-color);
            & svg {
                height: var(--cg-nt-menu-icon-size);
                width: var(--cg-nt-menu-icon-size);
            }
        }
    }
    & .menu-item.selected:not(.is-label):not(.is-disabled) {
        background-color: var(--cg-nt-menu-item-bg-color-hover);
    }
}
```

Which will allow you to create art like so:
![CleanShot 2024-07-12 at 17 39 04@2x](https://github.com/user-attachments/assets/bc9e8c70-9d9a-4bb0-ba9a-ef4043640fe6)

**Reply by FelipeRearden** [reply]
Now I can read the CSS and understand it !!!

Amazing @chrisgurney , so happy to be able to create menus the way that I wanted 🙏

Thank you very much for making this possible 💙

**Reply by chrisgurney** [reply]
@FelipeRearden It's live now in [1.9.0-beta-4](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.9.0-beta-4)

**Reply by FelipeRearden** [reply]
Working absolutely perfect on version `beta-4`. This amazing Chris, the style settings makes easy for the all users.

Thanks again for your interest on this subject. Everyday NTB becomes more important to my workflow.

I wish you a fantastic day ☀️

**Comment by FelipeRearden** [comment]
Just to let you know @chrisgurney that the update from `1.8.6` to `1.9.2` in my main has been perfect, no issues so far. 

I am creating my menus and applying the`menu-toolbar-item` selectors in my main vault and all good 🙏
- I was able to do exactly what I neeed 🙏

I am gonna continue here and I let you know if something occurs on `1.9.2` 

I wish you a fantastic day ☀️

**Reply by chrisgurney** [reply]
Appreciate it as always @FelipeRearden!

A couple things to check out:

- I had to adjust the `focus` command highlighting (CSS) in order to fix a separate bug.
- Menu (and sub-menu) positioning.

**Reply by FelipeRearden** [reply]
> A couple things to check out:

Thank you very much for the info @chrisgurney  , all good with my CSS snippets from my end :) 

> 19.8 -> Fixed an issue where the previously-selected command wasn't showing in an item's settings.

And thanks for the fix on `1.9.8` I was writting the BR and I saw the new release. I am running late since last week and I am happy to see that you could reproduce on your end.

I am using NTB daily since version 1.9.2 and everything is running flawless so far, The new `Menu` type really change the way that I am using NTB. ☀️
